### PR TITLE
Feat/wrapper block type

### LINKED ISSUE #149: feat: create a type wrapper around starknet-rs block (newtype)
Current type format mappings:

image

```[tasklist]
### Tasks
- [x] rename `types.rs` into `models.rs`
- [x] create a wrapper struct around `MaybePendingBlockWithTxHashes`
- [x] create a wrapper struct around `MaybePendingBlockWithTxs`
- [x] create a trait called ConvertibleEthBlock
- [x] create a function called `into_eth_block` in this trait, currently, it'll take on argument, a dyn KakarotClient
```

### LINKED ISSUE #151: refactor: starknet_block_to_eth_block using type wrapper impl.

### PR DESCRIPTION
**Pull request by Eikix**
Time spent on this PR:

Closes #149 
Closes #151 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

### CODE REVIEW

File: crates/core/src/client/mod.rs [hunk 1]
```diff
@@ -170,15 +177,17 @@ impl KakarotClient for KakarotClientImpl {
...
+            let starknet_block = BlockWithTxs::new(block);
+            let block = starknet_block.to_eth_block(self).await?;
+            return Ok(block);
         } else {
-            MaybePendingStarknetBlock::BlockWithTxHashes(
-                self.client.get_block_with_tx_hashes(block_id).await?,
-            )
+            let block = self.inner.get_block_with_tx_hashes(block_id).await?;
+            let starknet_block = BlockWithTxHashes::new(block);
+            let block = starknet_block.to_eth_block(self).await?;
```
Comment (LucasLvy) [inline crates/core/src/client/mod.rs:188]:
```suggestion
            starknet_block.to_eth_block(self).await
```

File: crates/core/src/client/mod.rs [hunk 2]
```diff
@@ -170,15 +177,17 @@ impl KakarotClient for KakarotClientImpl {
...
+            return Ok(block);
```
Comment (LucasLvy) [inline crates/core/src/client/mod.rs:189]:
```suggestion
```

File: crates/core/src/client/models.rs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+        //TODO: Fetch real data
+        let nonce: Option<H64> = Some(H64::zero());
+
+        //TODO: Fetch real data
+        let size: Option<U256> = Some(U256::from(1_000_000));
+
+        // Bloom is a byte array of length 256
+        let logs_bloom = Bloom::default();
+        let extra_data = Bytes::from(b"0x00");
+        //TODO: Fetch real data
```
Comment (LucasLvy) [inline crates/core/src/client/models.rs:67]: format

**Review by LucasLvy — DISMISSED** [review]
bof

File: crates/core/src/client/models.rs (hunk 3 above)
Comment (Eikix) [reply crates/core/src/client/models.rs]: done

File: crates/core/src/client/constants.rs [hunk 4]
```diff
@@ -1,13 +1,12 @@
 pub const CHAIN_ID: u64 = 1_263_227_476;
 
-pub const STARKNET_NATIVE_TOKEN: &str =
-    "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7";
+pub const STARKNET_NATIVE_TOKEN: &str = "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7";
 
 pub mod selectors {
-    use starknet::{core::types::FieldElement, macros::selector};
+    use starknet::core::types::FieldElement;
```
Comment (jobez) [inline crates/core/src/client/constants.rs:6]: interesting that the way that doesn't share structure is preferred!

Comment (Eikix) [reply crates/core/src/client/constants.rs:6]:
```toml
edition = "2021"
newline_style = "Unix"
use_field_init_shorthand = true
use_small_heuristics = "Max"
use_try_shorthand = true
max_width = 120

# Unstable features below
unstable_features = true
version = "Two"
comment_width = 100
format_code_in_doc_comments = true
format_macro_bodies = true
format_macro_matchers = true
format_strings = true
imports_granularity = "Module"
group_imports = "StdExternalCrate"
normalize_comments = true
normalize_doc_attributes = true
wrap_comments = true

```

**Review by greged93 — COMMENTED** [review]
Suggestion to remove the huge matches by implementing util functions on the wrapper.

File: crates/core/src/models/mod.rs [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    pub contract_address: Address,
+    pub token_balance: Option<U256>,
+    pub error: Option<String>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TokenBalances {
+    pub address: Address,
+    pub token_balances: Vec<TokenBalance>,
+}
```
[the remark this answers is not in the document]
Comment (Eikix) [reply crates/core/src/models/mod.rs:26]: sure, can we create an issue?

File: crates/core/Cargo.toml [hunk 6]
```diff
@@ -6,21 +6,22 @@ description.workspace = true
 homepage.workspace = true
 
 [dependencies]
-async-trait = "0.1.58"
+async-trait.workspace = true
```
Comment (jobez) [inline crates/core/Cargo.toml:9]: is the rule for workspaces something like:

- any dependency that is shared across crates in a project with the same parameterization -> workspace?

File: crates/core/src/models/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (greged93) [reply crates/core/src/models/mod.rs:24]: [Done ](https://github.com/kkrt-labs/kakarot-rpc/issues/169)

File: crates/core/Cargo.toml (hunk 6 above)
Comment (Eikix) [reply crates/core/Cargo.toml:9]: yes

File: crates/core/src/client/helpers.rs [hunk 7]
```diff
@@ -621,210 +595,73 @@ mod tests {
...
-            .unwrap(),
-            FieldElement::from_hex_be(
-                "0000000000000000000000000000000000000000000000000000000000000002",
-            )
-            .unwrap(),
-            FieldElement::from_hex_be(
-                "00000000000000000000000000000000000000000000000000000000000fffff",
-            )
-            .unwrap(),
+            FieldElement::from_hex_be("0000000000000000000000000000000000000000000000000000000000000002").unwrap(),
```
Comment (jobez) [inline crates/core/src/client/helpers.rs:598]: can we add a comment of where this data is sourced from?

can be a separate issue. relates to the question i asked in telegram around the mock data i have for starknet event keys.

File: crates/core/src/models/mod.rs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+/// Implement getters for fields that are present in Starknet Blocks, both in pending and validated
+/// state. For example, `parent_hash` is present in both `PendingBlock` and `Block`.
+macro_rules! implement_starknet_block_getters {
```
Comment (jobez) [inline crates/core/src/models/mod.rs:28]: this is pretty cool!

**Review by jobez — COMMENTED** [review]
my sense is that we should consolidate mock blocks/transactions/events somewhere, with an explanation of how we are sourcing them. 

then we can parameterize our converting tests with them.

(outside the scope of this pr, can create issue)

wdyt?

File: crates/core/src/client/helpers.rs (hunk 7 above)
Comment (Eikix) [reply crates/core/src/client/helpers.rs:598]: yes agreed

Comment (Eikix) [reply crates/core/src/client/helpers.rs:598]: we should ban hardcoded values that arent comprehensive, it triggers me (i wrote this i think) smh

**Pull request merged**